guacamole-common-js/src/main/webapp/modules/Keyboard.js

```diff
@@ -984,8 +1003,15 @@ Guacamole.Keyboard = function(element) {
             // Release specific key if known
             var keysym = first.keysym;
             if (keysym) {
+
+                // Press CapsLock first if Keyup event is alone.
+                // https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent#Special_cases
+                if (keysym === 0xFFE5)
+                    guac_keyboard.press(keysym);
```

Comment:
Given the following events, happening in order:

1. `keydown` Caps Lock
2. `keyup` Caps Lock

due to a single key press, would this then result in the following series of Guacamole keyboard events:

1. `keydown` Caps Lock (due to original `keydown`)
2. `keyup` Caps Lock (due to unreliable `keydown`)
3. `keydown` Caps Lock (due to unreliable `keyup`)
4. `keyup` Caps Lock (due to original `keyup`)

effectively doubling things on the remote end?

Reply:
You are correct; looking for the Keydown and Keyup events have no effect:
```
3.key,5.65509,1.1;	18	15:08:08.502
3.key,5.65509,1.0;	18	15:08:08.503
3.key,5.65509,1.1;	18	15:08:08.513
3.key,5.65509,1.0;	18	15:08:08.513
```
This leaves relying only on assuming Mac OS is quirky; which may be true until the behavior is 
fixed in the browser(s) and we'd have to handle that later.. Are we OK with that though.

Reply:
Generally, yes, that seems the only safe option - always releasing after handling keydown for Caps Lock, and then leaving the keyup behavior unmodified. There's no need to inspect the keyboard object's event queue to determine whether a keyup event is already present, since if so that event will be handled immediately after the keydown handling is complete, and the keyup event will simply be ignored if the corresponding key has already been marked as released.

Alternatively, if you want to browser sniff the quirk, I think that would be fine as well given the docs you've cited, but we should look to avoid repeating the sniff and to make the representation of the quirk(s) more semantic. From the iterations so far, it looks like this may even be as simple as just adding a call to `release()` if Caps Lock is known to be unreliable based on the platform test.